### PR TITLE
Improved error message when an object with same name exists and both …

### PR DESCRIPTION
**Pull request by prashantgolash**
…name & generateName are part of objectMeta spec

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
If both "name" and "generateName" are present in objectMeta YAML spec, and an object with the same "name" already exists, the error thrown during creation is that of ServerTimeout. Essentially, "generateName" is ignored if "name" exists in the spec, so ideally the error shouldn't be related to "generateName" (i.e. timeout). This PR intends to solve that by setting "generateName" to empty if the "name" is already present.

**Which issue(s) this PR fixes**:

Fixes # 32220

**Special notes for your reviewer**:
This PR is carry forward work of the previous PR discussed here - [61642](https://github.com/kubernetes/kubernetes/pull/61642). I put some of my comments in this note - [32220](https://github.com/kubernetes/kubernetes/issues/32220).

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by prashantgolash** [comment]
@neolit123  - sorry for some noise here. I have to close the earlier PR which you had taken a look due to license issue. I have incorporated your comments from earlier PR and submitted this new one. Please take a look when you get a chance.

**Review by neolit123 — COMMENTED** [review]
/ok-to-test
/priority backlog

thanks for the update.
try pinging #sig-api-machinery on slack if you are not getting a review after a week or so.
thanks.

File: staging/src/k8s.io/apiserver/pkg/registry/rest/create.go [hunk 1]
```diff
@@ -90,6 +90,8 @@ func BeforeCreate(strategy RESTCreateStrategy, ctx context.Context, obj runtime.
 	FillObjectMetaSystemFields(objectMeta)
 	if len(objectMeta.GetGenerateName()) > 0 && len(objectMeta.GetName()) == 0 {
 		objectMeta.SetName(strategy.GenerateName(objectMeta.GetGenerateName()))
+	} else {
+		objectMeta.SetGenerateName("")
```
Comment (yue9944882) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/create.go]: the change breaks backward-compatibility. e.g. for now we can create an object w/ non-empty `generateName` and non-empty `name`, when we want to override the name w/ one generated at client-side.

**Comment by fedebongio** [comment]
/assign @lavalamp 
(it might be a couple of weeks until we can review properly this one, due to KubeCon)

File: staging/src/k8s.io/apiserver/pkg/registry/rest/create.go (hunk 1 above)
Comment (prashantgolash) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/create.go]: @yue9944882 - Thanks for taking a look. Can you please tell me a scenario where we override name with generateName when both of them are non-empty? AFAIk, generateName has no bearing if the name is also present in objectMeta spec, but I may be wrong.

**Comment by prashantgolash** [comment]
> /assign @lavalamp
> (it might be a couple of weeks until we can review properly this one, due to KubeCon)
Thanks for your help !!!

**Comment by prashantgolash** [comment]
@lavalamp  - Do you have some time to review this.

**Pull request closed without merging**